NDTensors/src/lib/TypeParameterAccessors/src/set_parameters.jl

```diff
@@ -0,0 +1,81 @@
+"""
+    set_parameters(type::DataType, parameters::Tuple)
+
+Set the parameters of DataType `type` to the tuple `paramters`.
+"""
+function set_parameters(type::DataType, parameters::Tuple)
+  return unspecify_parameters(type){parameters...}
+end
+
+"""
+    set_parameters(type::UnionAll, parameters::Tuple)
+
+Set the parameters of UnionAll `type` to `paramters`.
+"""
+Base.@assume_effects :foldable function set_parameters(type::UnionAll, parameters::Tuple)
+  return Base.rewrap_unionall(set_parameters(Base.unwrap_unionall(type), parameters), type)
+end
+
+"""
+    set_parameters(type)
+
+If set_parameters is called with no `parameters` it returns the input unchanged.
+"""
+set_parameters(type) = type
+
+"""
+    set_parameter(type::Type, pos::Int, val)
+
+Set the parameter of the Type `type` in the position `pos` with the value `val`
+"""
+Base.@assume_effects :foldable function set_parameter(type::Type, pos::Int, val)
```

Comment:
We may have to use `Compat.@assume_effects` (https://github.com/JuliaLang/Compat.jl?tab=readme-ov-file#supported-features) to support Julia versions below 1.8 (https://docs.julialang.org/en/v1/base/base/#Base.@assume_effects).

Reply:
I converted to `Compat.@assume_effects` and the code compiles but a big issue is that something like `set_parameter(Array, (Float32, 2))` is not type stable for julia versions which are below 1.8. I tried both 1.6 and 1.7 and both fail.

Reply:
I don't think we need to worry about type stability in older versions of Julia, if users want better performance from improved type stability they can use newer versions.